### ISSUE TITLE
Difficulty reading updated scene information in a script

### ISSUE THREAD

**Issue by sqward** [JacquesLucke/animation_nodes#1250]
I have an export script which I execute in Blender 2.81 in headless mode (on commandline). It's meant to export and animation I've created using Animation Nodes. My script worked fine in 2.79. In 2.81 my object animation is not updated: for every scene.frame_set() position, rotatino etc. is the same.

Can you help me get to the bottom of the problem? I must be missing something that was introduced in Blender 2.8?

Regards

**Comment by OmarEmaraDev** [comment]
This seems like the issue reported in #1171, is this the issue you are having?

**Comment by sqward** [comment]
I've tried to render my animation and it seems to behave as expected.

**Comment by OmarEmaraDev** [comment]
So the issue only occur when you are rendering from the command line?

**Comment by sqward** [comment]
I'm not even rendering. My aim is to export animation data. So for every frame I captura object transform and store it. But I don't render anything.

**Comment by OmarEmaraDev** [comment]
So you have a node tree that sets the transformation of the object, and then you run a script that access the transformation of the object?

**Comment by sqward** [comment]
That's correct.

**Comment by OmarEmaraDev** [comment]
Can you share a minimal file that reproduces the issue?

**Comment by sqward** [comment]
I will prepare one. Thanks for prompt response.

**Comment by sqward** [comment]
Hi and sorry for delay.

My toy export script looks like this:

```python
import bpy

def printObjectPosition(obj):
    print(obj.name, obj.matrix_world.to_translation())

def convertAnimation(scene):
    objs = scene.objects
    scene.frame_set(scene.frame_start)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
    scene.frame_set(scene.frame_start+100)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
  
convertAnimation(bpy.data.scenes['Scene'])
```

and blender invocation looks like this:

>$ ../sv2017/src/Tools/blender-2.81a-windows64/blender.exe -noaudio -y --background an_test.blend -P export_example.py
Blender 2.81 (sub 16) (hash f1aa4d18d49d built 2019-12-04 14:30:40)
Read prefs: C:\Users\sqward\AppData\Roaming\Blender Foundation\Blender\2.81\config\userpref.blend
found bundled python: C:\Work\devel\sv2017\src\Tools\blender-2.81a-windows64\2.81\python
Read blend: C:\Work\devel\animation_nodes_export\an_test.blend
Registered Animation Nodes
Plane <Vector (-0.5003, 1.7544, 0.0000)>
Plane.001 <Vector (4.2274, 0.1166, 0.0000)>
Plane <Vector (-0.5003, 1.7544, 0.0000)>
Plane.001 <Vector (0.5983, -0.7396, 0.0000)>

My attached blend file contains two objects: "Plane" which is driven by AN and "Plane.001" which moved via key frames.

Note how position of "Plane.001" changes between ```scene.frame_set(scene.frame_start)``` and ```scene.frame_set(scene.frame_start+100)``` but not for "Plane".

This worked fine in 2.7x. Am I missing a trick in 2.8x?

[an_test.zip](https://github.com/JacquesLucke/animation_nodes/files/3947481/an_test.zip)

**Comment by OmarEmaraDev** [comment]
Animation Nodes needs to execute when you change the current frame in order for the object location to be updated. Animation Nodes can only see such change in the current frame after the script finish execution, so Auto Execution is not an option. To fix this, you just need to execute the node tree yourself when needed. Your script should be something like:

```python
import bpy

def printObjectPosition(obj):
    print(obj.name, obj.matrix_world.to_translation())

def convertAnimation(scene):
    objs = scene.objects
    scene.frame_set(scene.frame_start)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
    scene.frame_set(scene.frame_start+100)
    bpy.data.node_groups['NodeTree'].execute() # Execute node tree after change frame.
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
  
convertAnimation(bpy.data.scenes['Scene'])
```

Makes sense?

**Comment by sqward** [comment]
Yes. That makes perfect sense. I'm only curious why this wasn't necessary
in 2.79? Regards.

**Comment by OmarEmaraDev** [comment]
Blender 2.8 removed the ALWAYS event handler we used in 2.7x to implement Auto Execution, so now we use an infinite timer. I don't know exactly why this happens, but that must be the cause of this.

**Comment by sqward** [comment]
Thanks. But if my node uses "frame changed" handler then shouldn't that
work? I never use ALWAYS for updates as this drains the CPU a lot.

**Comment by OmarEmaraDev** [comment]
If you enable Auto Execution, Animation Nodes is always executing as many times as possible, even if the *Always* option is disabled. Animation Nodes does not use a proper trigger system for auto execution. For instance, when you enable *Frame Changed*, in the background, the frame value is retrieved and stored as a string, then it is compared to the previously stored value, if Animation Nodes notices that the value changed, it then executes the node tree.

To summarize, if you disable the *Always* option, Animation Nodes is still always executing an internal trigger mechanism.

**Comment by sqward** [comment]
Ok, thanks. I'll try your solution when I'll get back home. Regards.

**Comment by sqward** [comment]
Hi and sorry for delay.

My toy export script looks like this:

```python
import bpy

def printObjectPosition(obj):
    print(obj.name, obj.matrix_world.to_translation())

def convertAnimation(scene):
    objs = scene.objects
    scene.frame_set(scene.frame_start)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
    scene.frame_set(scene.frame_start+100)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
  
convertAnimation(bpy.data.scenes['Scene'])
```

and blender invocation looks like this:

>$ ../sv2017/src/Tools/blender-2.81a-windows64/blender.exe -noaudio -y --background an_test.blend -P export_example.py
Blender 2.81 (sub 16) (hash f1aa4d18d49d built 2019-12-04 14:30:40)
Read prefs: C:\Users\sqward\AppData\Roaming\Blender Foundation\Blender\2.81\config\userpref.blend
found bundled python: C:\Work\devel\sv2017\src\Tools\blender-2.81a-windows64\2.81\python
Read blend: C:\Work\devel\animation_nodes_export\an_test.blend
Registered Animation Nodes
Plane <Vector (-0.5003, 1.7544, 0.0000)>
Plane.001 <Vector (4.2274, 0.1166, 0.0000)>
Plane <Vector (-0.5003, 1.7544, 0.0000)>
Plane.001 <Vector (0.5983, -0.7396, 0.0000)>

My attached blend file contains two objects: "Plane" which is driven by AN and "Plane.001" which moved via key frames.

Note how position of "Plane.001" changes between ```scene.frame_set(scene.frame_start)``` and ```scene.frame_set(scene.frame_start+100)``` but not for "Plane".

This worked fine in 2.7x. Am I missing a trick in 2.8x?

[an_test.zip](https://github.com/JacquesLucke/animation_nodes/files/3947481/an_test.zip)

**Comment by sqward** [comment]
```bpy.data.node_groups['NodeTree'].execute()``` doesn't seem to help. Position of "Plane" object is not updated.  Did it work for you? Regards

**Comment by OmarEmaraDev** [comment]
Yes, it works with the blend file you sent.

**Comment by sqward** [comment]
Position of "Plane" stays the same after calling execute. Could that be platform difference?

```python
$ ../sv2017/src/Tools/blender-2.81a-windows64/blender.exe -noaudio -y --background an_test.blend -P export_example.py
Blender 2.81 (sub 16) (hash f1aa4d18d49d built 2019-12-04 14:30:40)
Read prefs: C:\Users\sqward\AppData\Roaming\Blender Foundation\Blender\2.81\config\userpref.blend
found bundled python: C:\Work\devel\sv2017\src\Tools\blender-2.81a-windows64\2.81\python
Read blend: C:\Work\devel\animation_nodes_export\an_test.blend
Registered Animation Nodes
Plane <Vector (0.3774, 1.7612, 0.0000)>
Plane.001 <Vector (4.2274, 0.1166, 0.0000)>
Plane <Vector (0.3774, 1.7612, 0.0000)>
Plane.001 <Vector (0.5983, -0.7396, 0.0000)>
```

**Comment by sqward** [comment]
It doesn't work for me from within Blender either:
![image](https://user-images.githubusercontent.com/627543/70667233-3fc0d500-1c68-11ea-941f-b8d082f74143.png)

**Comment by OmarEmaraDev** [comment]
Please note that Blender doesn't update matrices by default when they are changed, so you have to update them yourself. You can use the *Update Matrices* node for that.

![20191212-103129](https://user-images.githubusercontent.com/25300994/70695431-5c8df480-1cca-11ea-9404-a4cc5b252f56.png)

**Comment by sqward** [comment]
"Update Object Matrices" doesn't help. Are we using different versions of the software? :)

Note how "Plane" position didn't change again:

![image](https://user-images.githubusercontent.com/627543/70745432-b074f980-1d1b-11ea-9b5e-3773f209b66b.png)

**Comment by OmarEmaraDev** [comment]
hmmmm, no. Can you update the matrices from the script itself. See:
https://stackoverflow.com/questions/13840418/force-matrix-world-to-be-recalculated-in-blender

Otherwise, I would consider this a bug as in #1171.

**Comment by sqward** [comment]
I see my ```object.location``` doesn't change so world_matrix would be the same.

**Comment by OmarEmaraDev** [comment]
In this case, I would need to investigate this further. Might take some time, sorry.

**Comment by OmarEmaraDev** [comment]
Fixed in #1294 and b1ac938.

**Comment by sqward** [comment]
Thanks! I tried to find a relevant build but had no luck. Is that available?

**Comment by OmarEmaraDev** [comment]
Should be available in https://animation-nodes.com/

**Comment by sqward** [comment]
Thanks. In that case the might be something wrong with the build:
![image](https://user-images.githubusercontent.com/627543/74685022-3b172080-51c5-11ea-8e91-8ef8c83289ca.png)

**Comment by OmarEmaraDev** [comment]
See #1240.

**Comment by sqward** [comment]
Thanks. I can run it now.

Sadly, the above toy script still doesn't work. I've tried all your earlier suggestions (except bpy.data.node_groups['NodeTree'].execute() which doesn't seem to work at all).

**Comment by OmarEmaraDev** [comment]
hmm, let me test this again.

**Comment by OmarEmaraDev** [comment]
It seems to be working fine for me:

```cmd
┌─[omar]──[~/Desktop]
└─╼ blender -noaudio -y --background an_test.blend -P export_example.py
Blender 2.83 (sub 4) (hash dcb905a58f59 built 2020-02-17 14:36:20)
Read prefs: /home/omar/.config/blender/2.83/config/userpref.blend
/run/user/1000/gvfs/ non-existent directory
Registered Animation Nodes
Read blend: /home/omar/Desktop/an_test.blend
Plane <Vector (0.3774, 1.7612, 0.0000)>
Plane.001 <Vector (4.2274, 0.1166, 0.0000)>
Plane <Vector (-2.3628, -1.3816, 0.0000)>
Plane.001 <Vector (0.5983, -0.7396, 0.0000)>

Blender quit
```

This was using your original toy script:

```
import bpy

def printObjectPosition(obj):
    print(obj.name, obj.matrix_world.to_translation())

def convertAnimation(scene):
    objs = scene.objects
    scene.frame_set(scene.frame_start)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
    scene.frame_set(scene.frame_start+100)
    printObjectPosition(objs["Plane"])   
    printObjectPosition(objs["Plane.001"])   
  
convertAnimation(bpy.data.scenes['Scene'])
```

We now execute automatically on `frame_set`, so you don't have to execute manually.

Can you test with 2.83? Just so that we have the same setup and can debug this better.

**Comment by sqward** [comment]
Ok, it seem to work on 2.83... But I had to remove the "Update Object Matrices" node. Is this what you'd expect? Regards

**Comment by OmarEmaraDev** [comment]
Not exactly. Does it fail with the `Update Object Matrices` node there?

**Comment by sqward** [comment]
Yes, It fails to update the positions.

**Comment by OmarEmaraDev** [comment]
I am not entirely sure why this is the case. It could be that the depsgraph at this point is not up to date. But look at the bright side, it works. :grinning:

**Comment by sqward** [comment]
Indeed. Thanks! :)
